lib/action_client/submittable_request.rb

```diff
@@ -8,7 +8,13 @@ def initialize(stack, env)
     def submit
       app = @stack.build(ActionClient::Applications::Net::HttpClient.new)
 
-      app.call(env)
+      status, headers, body = app.call(env)
+
+      Rack::Response.new(body, status, headers) do |response|
```

Comment:
I wonder if a subclass of `Rack::Response` would help with the parsing issues discussed in #14?

We could provide a `ActionClient::Response#parsed_body` method which would do the parsing the first time it was invoked, and could raise if the body was invalid. I think that would give callers more flexibility to handle errors.

If we raised for an invalid response body within the current middleware, from a caller's perspective the entire request would have failed.

Reply:
Thanks @georgebrock ! I like that idea. It's could help consumers recover from bad data.

What I'm unsure about is how it would affect the current proposed [`after_submit` implementation](https://github.com/thoughtbot/action_client/pull/3), since those blocks are executed sequentially, expect triplets, and are dependent on modifications made to the body by their predecessors (potentially including parsing).

Would it make sense to merge #3, #10, #11, and #14, and then reassess?

Reply:
That's a really good point: I hadn't considered that `after_submit` callbacks currently get the parsed body rather than the string.

Thinking about it though, maybe it would make sense for callbacks to also take a response object?

We could add convenience methods to the response class to make it easier to e.g. replace the body. In the future (when we know what common callbacks are like from experience) we might even end up with some nice short cuts, like `after_submit(:not_found, &:raise_for_status_code)`

Reply:
> Would it make sense to merge #3, #10, #11, and #14, and then reassess?

Yes! I think it would be good to get all of this merged and try it out again in some real apps.